simulation/navigator_2Dsim/sim2d.py

```diff
@@ -0,0 +1,264 @@
+        return posedot, twistdot
+
+    def wrench_cb(self, msg):
+        self.force = msg.wrench.force
+        self.torque = msg.wrench.torque
+        # Set desired force and torque as numpy array
+        self.des_force = np.array((self.force.x, self.force.y, self.torque.z))
+
+    def publish_odom(self, *args):
+        if self.last_odom is not None:
+            self.odompublisher.publish(self.last_odom)
+            # Absodom publishing not implemented in this version
+        # if self.last_absodom is not None:
+            # self.absodompublisher.publish(self.last_absodom)
+    '''
+    # This may be used later to implement publishing to /absodom
+
+    def enu_to_ecef(self, enu):
+        if self.last_enu is None:
+            return self.last_ecef
+
+        enu_vector = enu - self.last_enu[0]
+        ecef_vector = gps.enu_from_ecef_tf(self.last_ecef).T.dot(enu_vector)
+        ecef = ecef_vector + self.last_ecef
+
+        return ecef
+    '''
+
+    def pack_odom(self, pose, twist):
+        """
+        Converts a state vector into an Odometry message
+        with a given header timestamp.
+
+        """
+        msg = Odometry()
+        msg.header.stamp = rospy.Time.now()
+        msg.header.frame_id = self.world_frame_id
+        msg.child_frame_id = self.body_frame_id
+        msg.pose.pose.position.x, msg.pose.pose.position.y, msg.pose.pose.position.z = pose
+        # print(msg.pose.pose.position.z)
+        quat = trns.quaternion_from_euler(0, 0, pose[2])
+        msg.pose.pose.orientation = mil_tools.numpy_to_quaternion(quat)
+        msg.twist.twist.linear.x, msg.twist.twist.linear.y = twist[0:2]
+        msg.twist.twist.angular.z = twist[2]
+        return msg
+
+    def state_cb(self):
+        if rospy.Time.now() - self.last_state_sub_time < self.state_sub_max_prd:
+            return
+
+        self.last_state_sub_time = rospy.Time.now()
+
+        # self.last_ecef = self.enu_to_ecef(self.pose)
+        self.last_enu = self.pose
+        # print(self.last_enu)
+        self.last_odom = self.pack_odom(self.pose, self.twist)
+
+        # self.last_absodom = self.pack_odom(self.last_ecef, self.twist)
+
+if __name__ == '__main__':
+
+    # Check to see if additional command line argument was passed.
+    # Allows 2D sim to play bags on its own, bag must be in the same
+    # directory. Fix this later.
+
+    # coordinate conversion server
+    # Prep
+
+    navsim = Navsim(pose0=np.array(
+        [0, 0, np.pi / 2]))
+
+    # Setup the graphing if enabled. Only done once.
+    if navsim.gui == True:
+        plt.ion()
+
+        fig1 = plt.figure()
+        fig1.suptitle('State Evolution', fontsize=20)
+        fig1rows = 2
+        fig1cols = 4
+    # print(navsim.gui)
+    dt = .05
+    rospy.sleep(5)
+
+    while not rospy.is_shutdown():
+
+        # Increment sim
+        navsim.state_cb()
+        navsim.publish_odom()
+        navsim.wrenches.append(navsim.des_force)
+        navsim.step(dt, navsim.wrenches[-1])
```

Comment:
Looks like this will just run in an infinite loop with no sleep now... Again I recommend doing all state calculations in a ROS timer. Nothing needs to be or should be dependent on unpredictable timing events happening at the right time.